### PR TITLE
CVE-2019-11246: Clean links handling in cp's tar code

### PR DESCRIPTION
**Pull request by soltysh**
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:
This PR cleans up the tar code used in `kubectl cp`. 

**Special notes for your reviewer**:
/assign @tallclair 

**Does this PR introduce a user-facing change?**:
```release-note
Clean links handling in cp's tar code
```

### CODE REVIEW

File: pkg/kubectl/cmd/cp/cp.go [hunk 1]
```diff
@@ -431,52 +429,60 @@ func (o *CopyOptions) untarAll(reader io.Reader, destFile, prefix string) error
...
-			if err != nil {
-				return err
-			}
-			if exists {
-				outFileName = filepath.Join(outFileName, path.Base(clean(header.Name)))
-			}
+		// We need to ensure that the destination file is always within boundries
+		// of the destination directory. This prevents any kind of path traversal
+		// from within tar archive.
+		dir, file := filepath.Split(destFileName)
```
[the remark this answers is not in the document]
Comment (soltysh) [reply pkg/kubectl/cmd/cp/cp.go]: Based on https://golang.org/pkg/path/filepath/#Split it should be solid.

File: pkg/kubectl/cmd/cp/cp.go [hunk 2]
```diff
@@ -490,14 +496,32 @@ func (o *CopyOptions) untarAll(reader io.Reader, destFile, prefix string) error
 		}
 	}
 
-	if entrySeq == -1 {
-		//if no file was copied
-		errInfo := fmt.Sprintf("error: %s no such file or directory", prefix)
-		return errors.New(errInfo)
-	}
```
[the remark this answers is not in the document]
Comment (soltysh) [reply pkg/kubectl/cmd/cp/cp.go]: The entire code was introduced in https://github.com/kubernetes/kubernetes/pull/46762 to deal with copying a file into a dir, which is now much simpler. I've wanted to clean this so with this significant changes it was a good time to do it.

File: pkg/kubectl/cmd/cp/cp.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (tallclair) [reply pkg/kubectl/cmd/cp/cp.go]: Looking at the [implementation](https://golang.org/src/path/filepath/path.go?s=5432:5474#L185), it seems like the destination file would be included in the directory if the path end sin a slash, and the file would be "" I think.

I think it's fine, but might be worth adding a test case?

[the remark this answers is not in the document]
Comment (soltysh) [reply pkg/kubectl/cmd/cp/cp.go]: I've added directory test, but that's actually handled a few lines earlier in 
```
destFileName := path.Join(destDir, header.Name[len(prefix):])
```
`destFileName` will get cleaned (`path.Join` calls `Clean` internally) even if it ends with `/`.

**Comment by soltysh** [comment]
@tallclair added that one last test case and also fixed the verify failure, ptal

File: pkg/kubectl/cmd/cp/cp_test.go [hunk 3]
```diff
@@ -768,6 +738,184 @@ func TestValidate(t *testing.T) {
...
+			// if you just look at the target relative to the path it appears
+			// inside the destination directory, but if you actually follow each
+			// step of the path you end up outside the destination directory.
+			path:       "nested/back-link-first/back-link-second/back-link-term",
+			linkTarget: "",
+			expected:   "",
+		})
+
+	files = append(files,
+		file{ // Absolute file
```
Comment (tallclair) [inline pkg/kubectl/cmd/cp/cp_test.go]:
```suggestion
		file{ // Relative directory path with terminating /
```

**Review by tallclair — COMMENTED** [review]
lgtm, just fix the comment.

**Comment by soltysh** [comment]
@tallclair updated, ptal

**Comment by soltysh** [comment]
Now with fixed e2e.

**Comment by soltysh** [comment]
Self-tagging based on previous approval.

File: pkg/kubectl/cmd/cp/cp.go [hunk 4]
```diff
@@ -432,52 +430,60 @@ func (o *CopyOptions) untarAll(reader io.Reader, destFile, prefix string) error
...
+		// of the destination directory. This prevents any kind of path traversal
+		// from within tar archive.
+		dir, file := filepath.Split(destFileName)
+		evaledPath, err := filepath.EvalSymlinks(dir)
+		if err != nil {
+			return err
+		}
+		// For scrutiny we verify both the actual destination as well as we follow
+		// all the links that might lead outside of the destination directory.
+		if !isDestRelative(destDir, destFileName) || !isDestRelative(destDir, filepath.Join(evaledPath, file)) {
```
Comment (abursavich) [inline pkg/kubectl/cmd/cp/cp.go]: I haven't had enough time to go through everything to understand the intention here, but I think the symlink checking is probably more aggressive than it should be.  Picking up this change breaks existing scripts that copy files to temp directories on Mac.

Please be aware that Mac's default `readlink` doesn't support canonicalizing a path, so it's not as easy to portably fix on the user's side as you might think.

```
$ FILE=path/to/file
$ TMPD=$(mktemp -d)
$ kubectl cp --no-preserve "$NAMESPACE/$POD:$FILE" "$TMPD/$FILE" -c $CONTAINER
warning: link "/var/folders/h7/ydd9wql57gdf6883cj9779sm0zsnn7/T/tmp.3TVUBz8Mqe/path/to/file" is pointing to "" which is outside target destination, skipping

$ echo $TMPD
/var/folders/h7/ydd9wql57gdf6883cj9779sm0zsnn7/T/tmp.3TVUBz8Mqe
$ readlink -f $TMPD  # using GNU readlink
/private/var/folders/h7/ydd9wql57gdf6883cj9779sm0zsnn7/T/tmp.3TVUBz8Mqe
```

**Pull request merged**